File: read_internal.c
/* Low-level data manipulation functions. */

/* What follows is low-level accessor functions, in order to inspect
   in details the contents of SEXPs and VECSEXPs. The implementation
   details are in Rinternals.h, enclosed in #ifdef USE_RINTERNALS
   directives, making these opaque pointers when not turned on. */

/* Concerning VECSEXPs: */

/* Quotation from section 1.1.3 "The 'data'" of "R Internals" (R-ints.pdf):
   'This [i.e. vecsxp.truelength] is almost unused. The only current use is for
   hash tables of environments (VECSXPs), where length is the size of the table
   and truelength is the number of primary slots in use, and for the reference
   hash tables in serialization (VECSXPs),where truelength is the number of
   slots in use.' */

/* CHARSXPs are in fact VECSEXPs. Concerning encoding, quotation from section 1.1.2
   "Rest_of_header" of "R internals" (R-ints.pdf): 'As from R 2.5.0, bits 2 and 3 for
   a CHARSXP are used to note that it is known to be in Latin-1 and UTF-8 respectively.
   (These are not usually set if it is also known to be in ASCII, since code does not
   need to know the charset to handle ASCII strings. From R 2.8.0 it is guaranteed that
   they will not be set for CHARSXPs created by R itself.) As from R 2.8.0 bit 5 is used
   to indicate that a CHARSXP is hashed by its address, that is NA STRING or in the
   CHARSXP cache. */


/**  Returns the attributes of an R value.
  *
  *  @param sexp An R value.
  *  @return The attributes of this value, as a pairlist.
  */
CAMLprim value ocamlr_inspect_attributes (value sexp) {
  return(Val_sexp(ATTRIB(Sexp_val(sexp))));
}


/**  Returns the length of a VECSEXP.
  *
  *  @param vecsexp An R vector value.
  *  @return Its length, i.e. the number of elements.
  */
CAMLprim value ocamlr_inspect_vecsxp_length (value vecsexp) {
  return(Val_int(LENGTH(Vecsexp_val(vecsexp))));
}



/* Concerning various types of SEXPs: */


/**  Returns the offset of an R primitive function.
  *
  *  @note: This function bypasses the provided API, and requires the
  *         use of the #define USE_RINTERNALS directive.
  *
  *  @param sexp An R value which is a primitive function.
  *  @return Its offset in the R table of primitives.
  */
CAMLprim value ocamlr_inspect_primsxp_offset (value sexp) {
  return(Val_int(Sexp_val(sexp)->u.primsxp.offset));
}


/**  Returns the name of an R symbol.
  *
  *  @param sexp An R value of sexptype SYMSXP.
  *  @return The name of the R symbol.
  */
CAMLprim value ocamlr_inspect_symsxp_pname (value sexp) {
  return(Val_sexp(PRINTNAME(Sexp_val(sexp))));
}


/**  Returns the value of a symbol.
  *
  *  @param sexp An R value of sexptype SYMSXP.
  *  @return The value of the R symbol.
  */
CAMLprim value ocamlr_inspect_symsxp_value (value sexp) {
  return(Val_sexp(SYMVALUE(Sexp_val(sexp))));
}


/**  Returns the internal sexp of a symbol.
  *
  *  @param sexp An R value of sexptype SYMSXP.
  *  @return The internal value of a symbol.
  */
CAMLprim value ocamlr_inspect_symsxp_internal (value sexp) {
  return(Val_sexp(INTERNAL(Sexp_val(sexp))));
}


/**  Returns the head element of a pairlist.
  *
  *  @param sexp An R pairlist.
  *  @return The head element of the R pairlist.
  */
CAMLprim value ocamlr_inspect_listsxp_carval (value sexp) {
  return(Val_sexp(CAR(Sexp_val(sexp))));
}


/**  Returns the tail pairlist of a pairlist.
  *
  *  @param sexp An R pairlist.
  *  @return The tail pairlist of the R pairlist.
  */
CAMLprim value ocamlr_inspect_listsxp_cdrval (value sexp) {
  return(Val_sexp(CDR(Sexp_val(sexp))));
}


/**  Returns the tag value of the head element of a pairlist.
  *
  *  @param sexp An R pairlist.
  *  @return The tag of the head of the R pairlist.
  */
CAMLprim value ocamlr_inspect_listsxp_tagval (value sexp) {
  return(Val_sexp(TAG(Sexp_val(sexp))));
}


/**  Returns the frame of an environment.
  *
  *  @param sexp An R environment.
  *  @return The frame of the R environment.
  */
CAMLprim value ocamlr_inspect_envsxp_frame (value sexp) {
  return(Val_sexp(FRAME(Sexp_val(sexp))));
}


/**  Returns the enclosing environment of an environment.
  *
  *  @param sexp An R environment.
  *  @return The enclosing environmnent of the R environment.
  */
CAMLprim value ocamlr_inspect_envsxp_enclos (value sexp) {
  return(Val_sexp(ENCLOS(Sexp_val(sexp))));
}


/**  Returns the hash table of an environment.
  *
  *  @param sexp An R environment.
  *  @return The hash table of the R environment.
  */
CAMLprim value ocamlr_inspect_envsxp_hashtab (value sexp) {
  return(Val_sexp(HASHTAB(Sexp_val(sexp))));
}


/**  Returns the list of formal arguments of a closure.
  *
  *  @param sexp An R closure.
  *  @return The list of formal arguments of the R closure.
  */
CAMLprim value ocamlr_inspect_closxp_formals (value sexp) {
  return(Val_sexp(FORMALS(Sexp_val(sexp))));
}


/**  Returns the body of a closure.
  *
  *  @param sexp An R closure.
  *  @return The body of the R closure.
  */
CAMLprim value ocamlr_inspect_closxp_body (value sexp) {
  return(Val_sexp(BODY(Sexp_val(sexp))));
}


/**  Returns the environment of a closure.
  *
  *  @param sexp An R closure.
  *  @return The environment of the R closure.
  */
CAMLprim value ocamlr_inspect_closxp_env (value sexp) {
  return(Val_sexp(CLOENV(Sexp_val(sexp))));
}


/**  Returns the value of a promise.
  *
  *  @note This function bypasses the provided API, and requires the
  *        use of the #define USE_RINTERNALS directive.
  *
  *  @param sexp An R promise.
  *  @return The value of the R promise.
  */
CAMLprim value ocamlr_inspect_promsxp_value (value sexp) {
  return(Val_sexp(Sexp_val(sexp)->u.promsxp.value));
}


/**  Returns the expression of a promise.
  *
  *  @note This function bypasses the provided R API, and requires the
  *        use of the #define USE_RINTERNALS directive.
  *
  *  @param sexp An R promise.
  *  @return The expression of the R promise.
  */
CAMLprim value ocamlr_inspect_promsxp_expr (value sexp) {
  return(Val_sexp(Sexp_val(sexp)->u.promsxp.expr));
}


/**  Returns the environment of a promise.
  *
  *  @note This function bypasses the provided R API, and requires the
  *        use of the #define USE_RINTERNALS directive.
  *
  *  @param sexp An R promise.
  *  @return The environment of the R promise.
  */
CAMLprim value ocamlr_inspect_promsxp_env (value sexp) {
  return(Val_sexp(Sexp_val(sexp)->u.promsxp.env));
}


/**  Returns an element of a logical vector.
  *
  *  @param lglsxp An R logical vector, of sexptype LGLSXP.
  *  @param offset An integer, offset in the R logical vector.
  *  @return The boolean at this offset in the R logical vector.
  */
CAMLprim value ocamlr_access_lgl_vecsxp (value lglsxp, value offset) {
  return(Val_bool(LOGICAL((int *) Vecsexp_val(lglsxp))[Int_val(offset)]));
}


/**  Returns an element of a vector of integers.
  *
  *  @param intsxp An R vector of integers, of sexptype INTSXP.
  *  @param offset An integer, offset in the R vector of integers.
  *  @return The integer at this offset in the R vector of integers.
  */
CAMLprim value ocamlr_access_int_vecsxp (value intsxp, value offset) {

  /* The R macro is #define INTEGER(x) ((int *) DATAPTR(x)).
     Should use Val_int, or int32s? More generally, the typing
     is here somewhat confusing (or confused)... Is offset an int? */

  return(Val_int(INTEGER((int *) Vecsexp_val(intsxp))[Int_val(offset)]));
}


/**  Returns an element of a vector of real numbers.
  *
  *  @param realsxp An R vector of real numbers.
  *  @param offset An integer, offset in the R vector of real numbers.
  *  @return The real number at this offset in the R vector of real numbers.
  */
CAMLprim value ocamlr_access_real_vecsxp (value realsxp, value offset) {
  return(caml_copy_double(REAL((double *) Vecsexp_val(realsxp))[Int_val(offset)]));
}


/**  Returns an element of a vector of strings.
  *
  *  @param strsxp An R vector of strings.
  *  @param offset An integer, offset in the R vector of strings.
  *  @return The string at this offset in the R vector of strings.
  */
CAMLprim value ocamlr_access_str_vecsxp (value strsxp, value offset) {

  /* Same comments as for r_access_int_vecsxp and for
     r_internal_string_of_charsxp. */

  return(caml_copy_string(CHAR(STRING_ELT((char **)
    Vecsexp_val(strsxp), (Int_val(offset))))));
}


/**  Returns an element of a vector of SEXPs.
  *
  *  r_access_sexp_vecsxp takes a vector of SEXPs as argument,
  *  and an offset, and returns the element at this offset.
  */

/* This function could also be called r_access_expr_vecsxp. */
/**  Returns an element of a vector of SEXPs.
  *
  *  @note This function is also used to retrieve an element in
  *        a vector of expressions, of sexptype EXPRSXP.
  *
  *  @param sexpsxp An R vector of SEXPs.
  *  @param offset An integer, offset in the R vector of SEXPs.
  *  @return The SEXP at this offset in the R vector of SEXPs.
  */
CAMLprim value ocamlr_access_sexp_vecsxp (value sexpsxp, value offset) {
  return(Val_sexp(VECTOR_ELT(Vecsexp_val(sexpsxp), Int_val(offset))));
}

File: read_internal.ml
(* Low-level data manipulation functions. *)

(* What follows is low-level accessor functions, in order to inspect
   in details the contents of SEXPs and VECSEXPs. *)


(**  Returns the attributes of a SEXP or VECSEXP
  *
  *  inspect_attributes takes a SEXP or a VECSEXP as
  *  arguments and returns its attributes, as a SEXP.
  *)

external inspect_attributes : sexp -> sexp = "inspect_attributes"


(**  Returns the length of a VECSEXP.
  *
  *  length_of_vecsxp takes a VECSEXP as argument and
  *  returns its length, i.e., the number of elements.
  *)

external length_of_vecsxp : 'a vecsxp -> int = "inspect_vecsxp_length"


(**  Returns the offset at the primitive can be found.
  *
  *  inspect_primsxp_offset take a SEXP denoting a function primitive
  *  as argument, and returns its offset in the table of primitives.
  *
  *  Note: This function bypasses the provided API, and requires the
  *  use of the #define USE_RINTERNALS directive.
  *)

external inspect_primsxp_offset  : builtin sxp -> int = "inspect_primsxp_offset"


(**  Returns the name of a symbol.
  *
  *  inspect_symsxp_pname take a SYMSXP as argument, and returns the
  *  SEXP containing its name.
  *)

external inspect_symsxp_pname    : 'a sym sxp -> sexp = "inspect_symsxp_pname"


(**  Returns the value of a symbol.
  *
  *  inspect_symsxp_value takes a SYMXSP as argument, and returns the
  *  SEXP containing its value.
  *)

external inspect_symsxp_value    : 'a sym sxp -> sexp = "inspect_symsxp_value"


(**  Returns the internal sexp of a symbol.
  *
  *  inspect_symsxp_internal takes a SYMSXP as argument, and returns the
  *  SEXP containing its internal value.
  *)

external inspect_symsxp_internal : 'a sym sxp -> sexp = "inspect_symsxp_internal"


(**  Returns the head element of a pairlist.
  *
  *  inspect_listsxp_carval takes a pairlist as argument, and
  *  returns the SEXP containing its head element.
  *)

external inspect_listsxp_carval  : 'a lisplist sxp -> sexp = "inspect_listsxp_carval"


(**  Returns the tail pairlist of a pairlist.
  *
  *  inspect_listsxp_cdrval takes a pairlist as argument,
  *  and returns its tail pairlist.
  *)

external inspect_listsxp_cdrval  : 'a lisplist sxp -> sexp = "inspect_listsxp_cdrval"


(**  Returns the tag value of the head element of a pairlist.
  *
  *  inspect_listsxp_tagval takes a pairlist as argument, and
  *  returns the tag value of the head element of the pairlist.
  *)

external inspect_listsxp_tagval  : 'a lisplist sxp -> sexp = "inspect_listsxp_tagval"


(**  Returns the frame of an evironment.
  *
  *  inspect_envsxp_frame takes an environment as argument,
  *  and returns the frame of this environment.
  *)

external inspect_envsxp_frame    : env sxp -> sexp = "inspect_envsxp_frame"


(**  Returns the enclosing environment of an environment.
  *
  *  inspect_envsxp_enclos takes an environment as argument,
  *  and returns its enclosing environment.
  *)

external inspect_envsxp_enclos   : env sxp -> sexp = "inspect_envsxp_enclos"


(**  Returns the hash table of an environment.
  *
  *  inspect_envsxp_hashtabl takes an environment as argument,
  *  and returns its hash table.
  *)

external inspect_envsxp_hashtab  : env sxp -> sexp = "inspect_envsxp_hashtab"


(**  Returns the list of formal arguments of a closure.
  *
  *  inspect_closxp_formals takes a closure as argument,
  *  and returns the list of its formal arguments.
  *)

external inspect_closxp_formals  : clos sxp -> sexp = "inspect_closxp_formals"


(**  Returns the body of a closure.
  *
  *  inspect_closxp_body takes a closure as argument,
  *  and returns the body of this closure.
  *)

external inspect_closxp_body     : clos sxp -> sexp = "inspect_closxp_body"


(**  Returns the environment of a closure.
  *
  *  inspect_closxp_env takes a closure as argument,
  *  and returns the environment of this closure.
  *)

external inspect_closxp_env      : clos sxp -> sexp = "inspect_closxp_env"


(**  Returns the value of a promise.
  *
  *  inspect_promsxp_value takes a promise as argument,
  *  and returns the value of this promise.
  *
  *  Note: This function bypasses the provided API, and requires the
  *  use of the #define USE_RINTERNALS directive.
  *)

external inspect_promsxp_value   : prom sxp -> sexp = "inspect_promsxp_value"


(**  Returns the expression of a promise.
  *
  *  inspect_promsxp_expr takes a promise as argument,
  *  and returns the value of this promise.
  *
  *  Note: This function bypasses the provided API, and requires the
  *  use of the #define USE_RINTERNALS directive.
  *)

external inspect_promsxp_expr    : prom sxp -> sexp = "inspect_promsxp_expr"


(**  Returns the environment of a promise.
  *
  *  inspect_promsxp_env takes a promise as argument,
  *  and returns the environment of this promise.
  *
  *  Note: This function bypasses the provided API, and requires the
  *  use of the #define USE_RINTERNALS directive.
  *)

external inspect_promsxp_env     : prom sxp -> sexp = "inspect_promsxp_env"


(**  Returns an element of a logical vector.
  *
  *  access_lgl_vecsxp takes a logical vector as argument,
  *  and an offset, and returns the element at this offset.
  *)

external access_lgl_vecsxp  : vec_lgl  sxp -> int -> bool         = "r_access_lgl_vecsxp"


(**  Returns an element of a vector of integers.
  *
  *  access_int_vecsxp takes a vector of integers as argument,
  *  and an offset, and returns the element at this offset.
  *)

external access_int_vecsxp  : vec_int  sxp -> int -> int          = "r_access_int_vecsxp"


(**  Returns an element of a vector of real numbers.
  *
  *  access_real_vecsxp takes a vector of integers as argument,
  *  and an offset, and returns the element at this offset.
  *)

external access_real_vecsxp : vec_real sxp -> int -> float        = "r_access_real_vecsxp"


(**  Returns an element of a vector of strings.
  *
  *  access_str_vecsxp takes a vector of strings as argument,
  *  and an offset, and returns the element at this offset.
  *)

external access_str_vecsxp  : vec_str  sxp -> int -> string       = "r_access_str_vecsxp"


(**  Returns an element of a vector of SEXPs.
  *
  *  access_sexp_vecsxp takes a vector of SEXPs as argument,
  *  and an offset, and returns the element at this offset.
  *)

external access_sexp_vecsxp : vec_sexp sxp -> int -> sexp         = "r_access_sexp_vecsxp"
